data_structure/trie_modern.cpp

```diff
@@ -0,0 +1,156 @@
+        if (word.size() == index) {
+            if (curr->isEndOfWord) {
+                curr->isEndOfWord = false;
+            }
+            if (hasChildren(curr)) {
+                return curr;
+            }
+            return nullptr;
+        }
+
+        size_t idx = word[index] - 'a';
+
+        // Throw a runtime error in case the user enters a word which is not
+        // present in the trie.
+        if (!curr->character[idx]) {
+            throw std::runtime_error(std::move(std::string("Word not found.")));
+        }
+
+        curr->character[idx] =
+            removeWordHelper(word, curr->character[idx], index + 1);
+
+        // This if condition checks if the node has some childern.
+        // The 1st if check, i.e. (curr->character[idx]) is checked specifically
+        // because if the older string is a prefix of some other string, then,
+        // there would be no need to check all 26 characters. Example- str1 =
+        // abbey, str2 = abbex and we want to delete string "abbey", then in
+        // this case, there would be no need to check all characters for the
+        // chars a,b,b.
+        if (curr->character[idx] || hasChildren(curr)) {
+            return curr;
+        }
+        return nullptr;
+    }
+
+ public:
+    // constructor to initialise the root of the trie.
+    Trie() : m_root(std::make_shared<TrieNode>()) {}
+
+    /**
+     * Insert a word into the trie.
+     * @param word which needs to be inserted into the string.
+     */
+    void insert(const std::string& word) {
+        auto curr = m_root;
+        for (char ch : word) {
+            size_t index = ch - 'a';
+
+            // if a node for current word is not already present in trie, create
+            // a new node for it.
+            if (!curr->character[index]) {
+                curr->character[index] = std::make_shared<TrieNode>();
+            }
+
+            curr = curr->character[index];
+        }
+        curr->isEndOfWord = true;
+    }
+
+    /**
+     * Search if a word is present in trie or not.
+     * @param word which is needed to be searched in the trie.
+     * @return if the word is found in the trie and isEndOfWord is set to true,
+     * then it returns @ true, else it returns @ false.
+     */
+    bool search(const std::string& word) {
+        auto curr = m_root;
+        for (char ch : word) {
+            size_t index = ch - 'a';
+
+            // if any node for a character is not found, then return that the
+            // word cannot be formed.
+            if (!curr->character[index]) {
+                return false;
+            }
+            curr = curr->character[index];
+        }
+        return curr->isEndOfWord;
+    }
+
+    // Function to remove the word which calls the helper function.
+    void removeWord(const std::string& word) {
+        m_root = removeWordHelper(word, m_root, 0);
+    }
+
+ private:
+    std::shared_ptr<TrieNode> m_root;
+};
+
+int main() {
```

Comment:
For the main, you can use a simple comment block like:
```
/**
 * Main function
 */
```

Reply:
Pushed with all the suggestions